### PR TITLE
Add YARD-style documentation.

### PR DESCRIPTION
**Pull request by leklund**
Reason for Change
===================
* Documentation!

Pull this branch and run 

```
bundle
yard
yard server
```
and you'll get a local yard server on which you can view the generated docs.

List of Changes
===============
* Documentation!

Risks
=====
* None

Checklist
=========
_Put an `x` in the boxes that apply. 
You can also fill these out after creating the PR._

- [x] I have NOT linked to any Jira tickets.
- [x] I have linked to all relevant reference issues or work requests
- [x] I have added tests that prove my fix is effective or that my feature works (N/A)
- [x] I have added or updated necessary documentation (if appropriate) 🤣 
- [x] Any dependent changes have been merged and published in downstream

Recommended Reviewers
=====================
@fastly/internal-engineering

### CODE REVIEW

File: lib/fastly_nsq.rb [hunk 1]
```diff
@@ -26,19 +37,37 @@ def logger
       @logger ||= Logger.new(nil)
     end
 
+    ##
+    # Configuration for FastlyNsq
```
Comment (set5think) [inline lib/fastly_nsq.rb:55]: Nice; we might be able to pull in some of the more in-depth examples in the readme into here.

Comment (leklund) [reply lib/fastly_nsq.rb:55]: oh. good idea.

File: lib/fastly_nsq.rb [hunk 2]
```diff
@@ -13,11 +13,25 @@ module FastlyNsq
   ConnectionFailed = Class.new(StandardError)
 
   class << self
+    # @return [String] NSQ Channel
     attr_accessor :channel
+
+    # @return [Proc] global preprocessor
     attr_accessor :preprocessor
+
+    # @return [Integer] maxium number of times an NSQ message will be attempted
```
Comment (lanej) [inline lib/fastly_nsq.rb]: `s/maxium/maximum/`

File: lib/fastly_nsq.rb [hunk 3]
```diff
@@ -13,11 +13,25 @@ module FastlyNsq
...
+
+    # @return [Logger]
     attr_writer :logger
 
+    ##
+    # Create a FastlyNsq::Listener
+    #
+    # @param topic [String] NSQ topic on which to listen
+    # @param processor [Proc] processor that will be +call+ed per message
+    # @param options [Hash] additional options
```
Comment (lanej) [inline lib/fastly_nsq.rb]: May wish to mention that the additional options are passed directly to the listener's constructor.

File: lib/fastly_nsq.rb [hunk 4]
```diff
@@ -26,19 +40,45 @@ def logger
...
+    ##
+    # Returns a new FastlyNsq::Manager or the memoized
+    # instance +@manager+.
+    # @return [FastlyNsq::Manager]
     def manager
       @manager ||= FastlyNsq::Manager.new
     end
 
+    ##
+    # Set a new manager
```
Comment (lanej) [inline lib/fastly_nsq.rb]: May want to mention that listeners are transferred to the new manager.

File: lib/fastly_nsq.rb [hunk 5]
```diff
@@ -26,19 +40,45 @@ def logger
...
+    # @param manager [FastlyNsq::Manager]
+    # @return [FastlyNsq::Manager]
     def manager=(manager)
       @manager&.transfer(manager)
       @manager = manager
     end
 
+    ##
+    # Return an array of NSQ lookupd http addresses
```
Comment (lanej) [inline lib/fastly_nsq.rb]: I'd call out that `NSQLOOKUPD_HTTP_ADDRESS` is used to source this information.

File: lib/fastly_nsq.rb (hunk 2 above)
Comment (lanej) [reply lib/fastly_nsq.rb]: Maybe also mention that `max_attempts == nil` means unlimited attempts?

File: lib/fastly_nsq/consumer.rb [hunk 6]
```diff
@@ -1,15 +1,81 @@
...
+  #   @see https://www.rubydoc.info/gems/nsq-ruby/Nsq%2FConsumer:pop_without_blocking Nsq::Consumer#pop_without_blocking
+  # @!method size
+  #   Delegated to +self.connection+
+  #   @return [Nsq::Consumer#size]
+  #   @see https://www.rubydoc.info/gems/nsq-ruby/Nsq%2FConsumer:size Nsq::Consumer#size
+  # @!method terminate
+  #   Delegated to +self.connection+
+  #   @return [Nsq::Consumer#terminate]
+  #   @see https://www.rubydoc.info/gems/nsq-ruby/Nsq%2FConsumer:terminate Nsq::Consumer#terminate
+  def_delegators :connection, :connected?, :pop, :pop_without_blocking, :size, :terminate
```
Comment (lanej) [inline lib/fastly_nsq/consumer.rb:59]: I would have expected yardoc to be smart enough to figure this out.

File: lib/fastly_nsq/consumer.rb [hunk 7]
```diff
@@ -28,6 +97,8 @@ def empty?
 
   attr_reader :tls_options
 
+  ##
+  # create a new Nsq::Consumer
```
Comment (lanej) [inline lib/fastly_nsq/consumer.rb]: I'm of the opinion that private method documentation should be optional.

File: lib/fastly_nsq/http.rb [hunk 8]
```diff
@@ -4,6 +4,16 @@
 require 'fastly_nsq/http/nsqd'
 require 'fastly_nsq/http/nsqlookupd'
 
+##
+# Adatper class for HTTP requests to NSQd
```
Comment (lanej) [inline lib/fastly_nsq/http.rb]: `s/Adatper/Adapter/g`

File: lib/fastly_nsq/listener.rb [hunk 9]
```diff
@@ -1,16 +1,80 @@
...
+  ##
+  # Create a FastlyNsq::Listener
+  #
+  # @param topic [String] NSQ topic on which to listen
+  # @param processor [Proc#call] Any object that responds to +call+. Each message will
+  #   be processed with +processor.call(FastlyNsq::Message.new(nsq_message))+.
+  #   The processor should return +true+ to indicate that processing is complete
+  #   and NSQ message can be finished. The processor is passed an instance of {FastlyNsq::Message}
+  #   so the provided Proc can optionally manage the message state using methods provided by {FastlyNsq::Message}.
+  # @param preprocessor [Proc#call] ANy object that responds to +call+. Similar to the processor
```
Comment (lanej) [inline lib/fastly_nsq/listener.rb]: `s/ANy/Any/`

File: lib/fastly_nsq/manager.rb [hunk 10]
```diff
@@ -1,11 +1,25 @@
 # frozen_string_literal: true
 
+##
+# Interface for tracking listeners.
```
Comment (lanej) [inline lib/fastly_nsq/manager.rb]: and managing the processing pool.

File: lib/fastly_nsq/message.rb [hunk 11]
```diff
@@ -2,14 +2,40 @@
...
 
-  attr_reader :managed, :nsq_message, :raw_body
+  # @return [Symbol] Message state. Returns +nil+ if message has not been requeued or finished.
+  attr_reader :managed
+
+  # @return [Nsq::Message]
+  # @see https://www.rubydoc.info/gems/nsq-ruby/Nsq/Message
+  attr_reader :nsq_message
+
+  # @return [String]Nsq::Message body
```
Comment (lanej) [inline lib/fastly_nsq/message.rb]: Missing a space.

File: lib/fastly_nsq/message.rb [hunk 12]
```diff
@@ -27,13 +53,18 @@ def body
...
   def finish
     return managed if managed
 
     @managed = :finished
     nsq_message.finish
   end
 
+  ##
+  # Requeue an NSQ Message
+  # @param timeout [Integer] timeout in milliseconds
```
Comment (lanej) [inline lib/fastly_nsq/message.rb:67]: I'm now realizing that `delay` would have been a better argument name.

File: lib/fastly_nsq/messenger.rb [hunk 13]
```diff
@@ -30,10 +49,16 @@ def producer_for(topic:)
     producer
   end
 
+  ##
+  # Hash of FastlyNsq::Producer objects
```
Comment (lanej) [inline lib/fastly_nsq/messenger.rb]: More accurately: Map of subscribed topics to `FastlyNsq::Producer`

File: lib/fastly_nsq/testing.rb [hunk 14]
```diff
@@ -1,6 +1,39 @@
 # frozen_string_literal: true
 
 module FastlyNsq
+  ##
```
Comment (lanej) [inline lib/fastly_nsq/testing.rb:4]: `FastlyNsq::Testing.message` is worth documenting or incorporating into the example.

File: lib/fastly_nsq/testing.rb [hunk 15]
```diff
@@ -57,7 +90,7 @@ def message(data:, meta: nil)
         FastlyNsq::Message.new(test_message)
       end
     end
-  end
+   end
```
Comment (lanej) [inline lib/fastly_nsq/testing.rb]: ?

File: lib/fastly_nsq/message.rb (hunk 12 above)
Comment (leklund) [reply lib/fastly_nsq/message.rb:67]: ah. true. Much more informative than `timeout`. I sense a refactoring in the future :)

File: lib/fastly_nsq/testing.rb (hunk 14 above)
Comment (leklund) [reply lib/fastly_nsq/testing.rb:4]: I added some brief docs and an example to `FastlyNsq::Testing.message` that I think should help explain usage of the test message.

**Review by set5think — APPROVED** [review]
Solid.

File: lib/fastly_nsq/testing.rb (hunk 14 above)
Comment (lanej) [reply lib/fastly_nsq/testing.rb:4]: Thanks!

**Pull request merged**